### PR TITLE
Azure outage postmortem

### PR DESCRIPTION
**Pull request by amy-jenn**
https://visual-meaning.atlassian.net/browse/SMP-2308

Postmortem on March outage for Azure

### CODE REVIEW

File: postmortems/2024-03-22_azure_dev_outage.md [hunk 1]
```diff
@@ -0,0 +1,110 @@
...
+* 2024/03/25 Monday 13:54 MP confirms that the app registration secret for the prod version of the app has expired and needs to be regenerated 
+* 2024/03/25 Monday 13:57 MP discovers his account for client's Azure env is blocked 
+* 2024/03/25 Monday 14:02 MP is reminded that the guest B2B account is unable to regenerate secrets, despite being an owner of the app registration for which the secret has expired
+* 2024/03/25 Monday 15:04 Development Team Sprint Planning ends 
+* 2024/03/25 Monday 15:19 MP calls IT Service Desk, account is blocked because it has expired and needs account extension from line manager for the password to be reset and the account reactivated. Expired due to original timing authorized for account.  
+* 2024/03/25 Monday 15:48 Line manager who must request the account extension is identified (normal LM is on long term leave) 
+* 2024/03/25 Monday 15:53 Notice added to client's Sharepoint indicating that app is down 
+* 2024/03/25 Monday 16:48 Build Pipelines and Azure Repo are discovered to be not visible to MP & AL who previous could see them – reason not clear at this stage
```
Comment (alightwing) [inline postmortems/2024-03-22_azure_dev_outage.md:26]: This was annoying and something that we did discover at the same time while we were debugging permissions, but I don't think it's relevant to the app outage?

File: postmortems/2024-03-22_azure_dev_outage.md [hunk 2]
```diff
@@ -0,0 +1,110 @@
...
+
+In addition to this, the first alert was received at 14:38 on the Friday, but only forwarded to MP & AL at 10:38 on Monday.  
+
+### Why was the secret allowed to expire? 
+
+1. No system reminder to renew secret 
+The client Azure system does not automatically send any notifications for impending expiry of secret – based on content in Azure forums it seems to be a common problem for Azure secrets beyond this client environment. There isn’t an obvious place for us to set these up for ourselves, but we have also never asked about it. We intend to look into some scripts that we could potentially run that would enable automated reminders for the secret expiry.  
+
+Given that one third party supplier sends an excel spreadsheet monthly of Service Principals expiring, we believe this must be a widespread problem.
```
Comment (alightwing) [inline postmortems/2024-03-22_azure_dev_outage.md:68]: I'm not sure this being a widespread problem is that relevant? Yes, Azure secret management is bad, but it's what we have to work with. It's worth having a note that this isn't a specific feature of the client Azure environment and is instead just a shortcoming of Azure in general, but in so far as this specific outage was concerned all that matters is that:

- Azure has no functionality to automatically alert app registration/service principal owners that an attached secret is about to expire.
- Therefore we need to make our own alerts or reminders, automated or otherwise.
- For this specific secret, we did not do this.

File: postmortems/2024-03-22_azure_dev_outage.md [hunk 3]
```diff
@@ -0,0 +1,110 @@
...
+
+Given that one third party supplier sends an excel spreadsheet monthly of Service Principals expiring, we believe this must be a widespread problem. 
+
+2. No manual reminder from client to renew secret 
+Third party supplier sends an excel spreadsheet monthly of Service Principal secrets expiring. The secret that expired was not listed in the spreadsheet that is sent monthly – 4 of our service principals are listed, but 4 of the other service principals for the app are not listed, and none of the app registrations are listed. The secret that expired and caused the lack of access was an app registration secret.  
+
+3. Inconsistent expiry timelines 
+Each of the different types of secrets have different expiry timelines, which makes tracking them harder. We can’t generically indicate that a year from now we need to redo this, so the dates need to be tracked separately and specifically rather than conceptually. Policy doc somewhere says things are supposed to expire 6 monthly, but most of the defaults have been set to annual expiry (e.g. cert is 3 months).
```
Comment (alightwing) [inline postmortems/2024-03-22_azure_dev_outage.md:74]: Which cert? The domain SSL cert?

File: postmortems/2024-03-22_azure_dev_outage.md [hunk 4]
```diff
@@ -0,0 +1,110 @@
...
+
+## Wonderings
+
+Why are B2B guest accounts added as app registration owners if they cannot exercise a primary privilege of app registration ownership? 
+
+Why did senior developers lose visibility of the pipelines and repos, and is this something that could happen again? Could similar loss of privilege occur without warning on other key areas? 
+
+Monitoring to detect the outage was ignored, it was only spotted when trying to access the service during working hours.
+
+There was clear and regular communication internally about the outage and progress to address it, and although initially stilted, eventually clear and regular communication with line manager (client side) on the matter.
```
Comment (alightwing) [inline postmortems/2024-03-22_azure_dev_outage.md:110]: I'm generally not a fan of closing with this section. I want to see the lessons learned and remediation actions at the end of a post-mortem, not open-ended questions.

**Review by alightwing — COMMENTED** [review]
I think this is a good first draft that captures all the detail - ironically, my main complaint is that it might be _too much_ detail.

We have two unrelated permissions issues in this post mortem, the app registration secret expiry and the developers having their Azure Devops access mysteriously downgraded.  Since this post mortem is about the app outage caused by the secret expiry, and this wasn't dependent on Azure Devops access in any way, I think including the detail about Azure Devops access makes the timeline especially a bit too hard to follow. I think that we should capture that detail somewhere, but not in this document.

A thing that's also missing is a consolidated list of learnings/actions. We do have some scattered throughout the doc, but if I'm someone external reading this I'd really like an executive summary of we're doing to prevent this from happening again in the future.

File: postmortems/2024-03-22_azure_dev_outage.md [hunk 5]
```diff
@@ -0,0 +1,110 @@
...
+2024 (all times UTC)
+* 2024/03/22 Friday 13:17 – app registration Secret expired 
+* 2024/03/22 Friday 14:38 – First request to authorisation server with a denial; generated an Azure Monitor Alert sent to Support Inbox 
+* 2024/03/25 Monday 10:38 – First Azure Monitor Alert forwarded to MP & AL (developers)
+* 2024/03/25 Monday 13:29 Development Team Sprint Planning session begins; Product Owner (PO), QA, and all developers are among attendees – this is only relevant as an indication that the team were in a meeting 
+* 2024/03/25 Monday 13:46 Internal user reports trying to access the site and being unable to log in 
+* 2024/03/25 Monday 13:47 Confirmed that second internal user is also not able to log in 
+* 2024/03/25 Monday 13:54 MP confirms that the app registration secret for the prod version of the app has expired and needs to be regenerated 
+* 2024/03/25 Monday 13:57 MP discovers his account for client's Azure env is blocked 
+* 2024/03/25 Monday 14:02 MP is reminded that the guest B2B account is unable to regenerate secrets, despite being an owner of the app registration for which the secret has expired
```
[the remark this answers is not in the document]
Comment (amy-jenn) [reply postmortems/2024-03-22_azure_dev_outage.md:21]: Martin knew it after the B2B account was added as an owner to the app registration for prod, which is why the client account was added to the app registration for dev - so yes, we have known about this for long enough to have already addressed it.

File: postmortems/2024-03-22_azure_dev_outage.md (hunk 1 above)
Comment (amy-jenn) [reply postmortems/2024-03-22_azure_dev_outage.md:26]: I think it contributes to the overall picture that there are unexpected things happening that pose a risk, and that demonstrate failure on both sides. Happy to remove though if you don't feel this is necessary. Shall I remove @alightwing?

File: postmortems/2024-03-22_azure_dev_outage.md (hunk 3 above)
Comment (amy-jenn) [reply postmortems/2024-03-22_azure_dev_outage.md:74]: I think so yes

File: postmortems/2024-03-22_azure_dev_outage.md (hunk 4 above)
Comment (amy-jenn) [reply postmortems/2024-03-22_azure_dev_outage.md:110]: Added some remediations

File: postmortems/2024-03-22_azure_dev_outage.md (hunk 3 above)
Comment (amy-jenn) [reply postmortems/2024-03-22_azure_dev_outage.md:69]: No changes based on this

File: postmortems/2024-03-22_azure_dev_outage.md (hunk 1 above)
Comment (amy-jenn) [reply postmortems/2024-03-22_azure_dev_outage.md]: Changes made

File: postmortems/2024-03-22_azure_dev_outage.md (hunk 5 above)
[the remark this answers is not in the document]
Comment (amy-jenn) [reply postmortems/2024-03-22_azure_dev_outage.md:21]: No changes made

**Review by alightwing — APPROVED** [review]
Thanks!
